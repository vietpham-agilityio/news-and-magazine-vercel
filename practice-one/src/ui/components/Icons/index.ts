export * from './User';
export * from './Plus';
export * from './Home';
export * from './File';
export * from './Close';
export * from './Circle';
export * from './Search';
export * from './SaveTag';
export * from './MailBox';
export * from './Twitter';
export * from './ChatBox';
export * from './Ellipse';
export * from './Schedule';
export * from './ArrowTop';
export * from './ThreeDots';
export * from './Instagram';
export * from './BurgerMenu';
export * from './ChevronsUp';
export * from './ArrowLeft';
export * from './ArrowRight';
export * from './PaperPlane';
export * from './ArrowBottom';
export * from './SaveTagSmall';