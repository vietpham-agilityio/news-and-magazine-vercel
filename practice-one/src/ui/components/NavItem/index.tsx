import { memo } from "react";

interface IProps {
  name: string;
  url?: string;
}

export const NavItem = memo(({ name, url }: IProps) => (
  <a href={url} className="flex items-center relative">
    <p className="text-sm font-medium text-dark-100 hover:text-primary-100">
      { name }
    </p>
  </a>
));

NavItem.displayName = "NavItem";
