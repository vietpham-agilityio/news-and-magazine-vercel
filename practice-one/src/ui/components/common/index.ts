export * from './Tag';
export * from './Menu';
export * from './Avatar';
export * from './Button';
export * from './BoxIcon';
export * from './SubMenu';
export * from './Typography';
export * from './Information';
export * from './TitleSection';
