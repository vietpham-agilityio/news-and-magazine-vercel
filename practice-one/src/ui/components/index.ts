// common
export * from '@/ui/components/common';

// shared
export * from '@/ui/components/Video';
export * from '@/ui/components/Brand';
export * from '@/ui/components/NavItem';
export * from '@/ui/components/Comment';
export * from '@/ui/components/Carousel';
export * from '@/ui/components/SearchBar';
export * from '@/ui/components/PaperPost';
export * from '@/ui/components/PaperVideo';
export * from '@/ui/components/BurgerMenu';
export * from '@/ui/components/Card/CardSingle';
export * from '@/ui/components/Skeleton/CardPost';
export * from '@/ui/components/Skeleton/DetailPost';
export * from '@/ui/components/Skeleton/CardVertical';
export { default as CardPost } from '@/ui/components/Card/CardPost';