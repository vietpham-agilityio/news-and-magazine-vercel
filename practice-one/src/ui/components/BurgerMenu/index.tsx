'use client';

import { useState, useEffect } from 'react';

// constants
import { flexItemCenter } from '@/constants';

// mocks
import { listNavItems } from '@/mocks';

// components
import { BoxIcon, Brand, NavItem, Typography } from '@/ui/components';
import Link from 'next/link';

// icons
import {
  ArrowBottom,
  ArrowRight,
  Close,
  BurgerMenu as BurgerMenuIcon,
} from '@/ui/components/Icons';

// types
import { FontWeight, IMenu, Size } from '@/types';
import { createPortal } from 'react-dom';

interface IProps {
  listCategories: IMenu[];
}

export const BurgerMenu = ({ listCategories }: IProps) => {
  const [isOpenMenu, setIsOpenMenu] = useState<boolean>(false);
  const [isOpenSubMenu, setIsOpenSubMenu] = useState<boolean>(false);

  const classNavItem = `w-full h-12 pl-2 rounded-xl hover:bg-white-90 ${flexItemCenter}`;

  const handleToggleMenu = () => setIsOpenMenu(prev => !prev);
  const handleCloseMenu = () => setIsOpenMenu(false);
  const handleCloseSubMenu = () => setIsOpenSubMenu(false);

  const handleToggleSubMenu = () => setIsOpenSubMenu(prev => !prev);

  const handleCloseAllMenu = () => (handleCloseMenu(), handleCloseSubMenu());

  useEffect(() => {
    isOpenMenu
      ? (document.body.style.overflow = 'hidden')
      : (document.body.style.overflow = 'auto');
  }, [isOpenMenu]);

  return (
    <div>
      <BoxIcon
        icon={<BurgerMenuIcon />}
        onClick={handleToggleMenu}
        additionalClasses="w-12"
      />
      {isOpenMenu &&
        createPortal(
          <div className="w-full h-full z-10 absolute px-5 pb-4 pt-6 left-0 top-0 bg-white-100 shadow-xl">
            <div className="flex relative">
              <Brand brandName="MEGA.news" url="/" />
              <div className="absolute -right-2 -top-2">
                <BoxIcon
                  icon={<Close />}
                  onClick={handleCloseMenu}
                  additionalClasses="w-12"
                />
              </div>
            </div>
            <div
              className="flex justify-between items-center mt-4 h-12 pl-2 pr-3 rounded-xl hover:bg-white-90"
              onClick={handleToggleSubMenu}
            >
              <div className={`${flexItemCenter}`}>
                {isOpenSubMenu && (
                  <div className="w-1 h-[11px] bg-primary-100 rounded-2xl mr-2" />
                )}
                <Typography
                  tag="p"
                  textSize={Size.SM}
                  weight={FontWeight.Medium}
                  additionalClasses="text-dark-100 hover:text-primary-100"
                >
                  Categories
                </Typography>
              </div>
              {isOpenSubMenu ? <ArrowBottom /> : <ArrowRight />}
            </div>

            {isOpenSubMenu && (
              <ul className="list-none mt-1.5 pl-5">
                {listCategories.map(category => {
                  const { id, name } = category;
                  return (
                    <li
                      key={id}
                      className="mb-2 last:mb-0"
                      onClick={handleCloseAllMenu}
                    >
                      <Link href={`/category/${id}`}>
                        <Typography
                          tag="p"
                          textSize={Size.XS}
                          additionalClasses="text-dark-75 cursor-pointer hover:text-secondary-100"
                        >
                          { name }
                        </Typography>
                      </Link>
                    </li>
                  );
                })}
              </ul>
            )}

            <div>
              {listNavItems.map(navItem => (
                <div key={navItem.id} className={classNavItem}>
                  <NavItem name={navItem.name} url={navItem.path} />
                </div>
              ))}
            </div>
          </div>,
          document.body
        )}
    </div>
  );
};
