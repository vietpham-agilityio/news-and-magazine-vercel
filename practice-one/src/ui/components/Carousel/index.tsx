'use client';

// hooks
import { useState, useEffect, useCallback } from 'react';

// components
import Image from 'next/image';
import { BoxIcon, Typography } from '@/ui/components';

// types
import { FontWeight, ISlideItem, Size } from '@/types';

// icons
import { ArrowLeft, ArrowRight, Circle, Ellipse } from '@/ui/components/Icons';

interface IProps {
  listSlider: ISlideItem[];
  autoPlay?: boolean;
  autoPlayTime?: number;
}

export const Carousel = ({
  listSlider,
  autoPlay = true,
  autoPlayTime = 3000,
}: IProps) => {
  const [currentIndex, setCurrentIndex] = useState(0);

  const goToPrevious = useCallback(() => {
    const isFirstSlide = currentIndex === 0;
    const newIndex = isFirstSlide ? listSlider.length - 1 : currentIndex - 1;
    setCurrentIndex(newIndex);
  }, [currentIndex, listSlider.length]);

  const goToNext = useCallback(() => {
    const isLastSlide = currentIndex === listSlider.length - 1;
    const newIndex = isLastSlide ? 0 : currentIndex + 1;
    setCurrentIndex(newIndex);
  }, [currentIndex, listSlider.length]);

  useEffect(() => {
    if (autoPlay) {
      const interval = setInterval(goToNext, autoPlayTime);
      return () => clearInterval(interval);
    }
  }, [currentIndex, autoPlay, autoPlayTime, goToNext]);

  return (
    <div className="relative flex flex-col items-center w-full h-full mx-auto overflow-hidden border-2 bg-black-100 border-dark-10 rounded-lg">
      <BoxIcon
        width="10"
        height="10"
        additionalClasses="absolute left-4 z-10 top-1/2 hover:opacity-80"
        icon={<ArrowLeft />}
        onClick={goToPrevious}
      />
      <div
        className="w-full flex transition-transform duration-500 xl:w-126.5 2xl:w-186"
        style={{ transform: `translateX(-${currentIndex * 100}%)` }}
      >
        {listSlider.map((slider, index) => {
          const { image, title, content } = slider;

          return (
            <div
              key={index}
              className="w-full h-113 flex-shrink-0 flex flex-col items-center relative"
            >
              <Image
                fill
                objectFit='cover'
                key={index}
                src={image}
                alt={`Slide ${index}`}
              />
              <div className="absolute left-5 bottom-3.5 p-4 bg-white-75 rounded-xl opacity-80 w-[93%] lg:w-[96%] xl:w-[92%] 2xl:w-[95%]">
                <Typography
                  tag="h2"
                  textSize={Size.LG}
                  weight={FontWeight.Normal}
                  additionalClasses="text-black-100 h-10 text-ellipsis overflow-hidden"
                >
                  {title}
                </Typography>
                <Typography
                  tag="h3"
                  textSize={Size.XSM}
                  weight={FontWeight.Light}
                  additionalClasses="text-dark-100 w-100 lg:w-full xl:w-75 2xl:w-[543px] h-10 leading-5 text-ellipsis overflow-hidden"
                >
                  {content}
                </Typography>
              </div>
            </div>
          );
        })}
      </div>
      <BoxIcon
        width="10"
        height="10"
        additionalClasses="absolute right-4 top-1/2 hover:opacity-80"
        icon={<ArrowRight />}
        onClick={goToNext}
      />

      <div className="flex absolute bottom-4 right-6">
        {listSlider.map((_, index) => (
          <button
            key={index}
            aria-label="slider-button"
            className="h-3 z-10 rounded-full w-6"
            onClick={() => setCurrentIndex(index)}
          >
            {index === currentIndex ? <Ellipse /> : <Circle />}
          </button>
        ))}
      </div>
    </div>
  );
};
