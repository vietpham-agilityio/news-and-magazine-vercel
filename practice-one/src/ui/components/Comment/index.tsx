// components
import { FontWeight, Size } from "@/types";
import { Typography } from "@/ui/components/common";

interface IProps {
  userName: string;
  content: string;
}

export const Comment = ({ userName, content }: IProps) => (
  <div className="bg-white-90 p-4 rounded-xl">
    <Typography tag="h2" textSize={Size.SM} additionalClasses="text-dark-100">
      { userName }
    </Typography>
    <Typography
      tag="h3"
      weight={FontWeight.Light}
      additionalClasses="text-p text-dark-75"
    >
      { content }
    </Typography>
  </div>
);
