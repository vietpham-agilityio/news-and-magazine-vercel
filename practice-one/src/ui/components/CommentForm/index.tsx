// components
import { Button, Input, TextArea } from '@/ui/components';

// icons
import { ChatBoxSmall } from '@/ui/components/Icons';

export const CommentForm = () => (
  <div>
    <div className="grid grid-cols-12 grid-rows-3 lg:gap-6">
      <div className="row-span-3 col-span-12 lg:col-span-6">
        <div className="mb-5 row-span-1">
          <Input label="Name" name="username" type="text" />
        </div>
        <div className="mb-5 row-span-1">
          <Input label="Email" name="email" type="email" />
        </div>
        <div className="mb-5 row-span-1">
          <Input label="Website" name="website" type="text" />
        </div>
      </div>
      <div className="row-span-3 col-span-12 lg:col-span-6">
        <TextArea label="Comment" placeholder="What do you think?" />
      </div>
    </div>
    <div className="flex w-full justify-end mt-7 mb-30 sm:mt-10 sm:mb-12 lg:mt-0">
      <Button
        isButtonLarge
        label="Send Comment"
        leftIcon={<ChatBoxSmall />}
        bgColor="bg-gradient-to-r from-sunSet-0 to-sunSet-100"
        textColor="text-white-100"
      />
    </div>
  </div>
);
