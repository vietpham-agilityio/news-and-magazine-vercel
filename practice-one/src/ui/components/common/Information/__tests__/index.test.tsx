import { render, screen } from '@testing-library/react';

// component
import { Information } from '@/ui/components';

let renderInformation: any;
const name = 'Bagari';
const date = '02/12/2024';

describe('Information component', () => {
  beforeEach(() => {
    renderInformation = render(<Information name={name} information={date} />);
  });

  afterEach(() => {
    renderInformation.unmount();
  });

  it('should render Information macth snapshot', () => {
    expect(renderInformation.asFragment()).toMatchSnapshot();
  });

  it('should render name value to macth snapshot', () => {
    const nameValue = screen.getByText(name);

    expect(nameValue).toBeInTheDocument();
  });

  it('should render information value to macth snapshot', () => {
    const informationValue = screen.getByText(date);

    expect(informationValue).toBeInTheDocument();
  });
});
