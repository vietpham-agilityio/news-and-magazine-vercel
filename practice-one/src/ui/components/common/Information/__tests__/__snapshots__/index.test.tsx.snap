// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Information component should render Information macth snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-between"
  >
    <p
      class="text-sm font-medium text-dark-100"
      role="heading"
    >
      Bagari
    </p>
    <p
      class="text-xs font-light text-dark-75"
      role="heading"
    >
      02/12/2024
    </p>
  </div>
</DocumentFragment>
`;
